File: api/pdf-server.js
const express = require('express');
const fs = require('fs');
const path = require('path');
const Handlebars = require('handlebars');
const puppeteer = require('puppeteer');
const { createClient } = require('@supabase/supabase-js');

const app = express();
const PORT = 3002;

const TEMPLATE_PATH = path.join(__dirname, '..', 'templates', 'resume-extended.html');

// Supabase 配置
const SUPABASE_URL = 'https://riieooizyhovmgvhpcxj.supabase.co';
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJpaWVvb2l6eWhvdm1ndmhwY3hqIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MTQwMjg1MCwiZXhwIjoyMDg2OTc4ODUwfQ.azMzZoioMnKKJwwwmaroxTxLnVYHMasfAxkW6lkdptk';

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

async function loadDataFromSupabase() {
  try {
    // 获取 profile
    const { data: profile, error: profileError } = await supabase
      .from('profiles')
      .select('*')
      .limit(1)
      .single();
    
    if (profileError) throw profileError;
    if (!profile) throw new Error('Profile not found');

    // 并行获取其他数据
    const [
      { data: education },
      { data: workExperience },
      { data: skillCategories },
      { data: projects }
    ] = await Promise.all([
      supabase.from('education').select('*').eq('profile_id', profile.id).order('sort_order'),
      supabase.from('work_experience').select('*').eq('profile_id', profile.id).order('sort_order'),
      supabase.from('skill_categories').select('*, skills(*)').eq('profile_id', profile.id).order('sort_order'),
      supabase.from('projects').select(`
        *,
        sub_projects(
          *,
          project_links(*)
        )
      `).eq('profile_id', profile.id).order('sort_order')
    ]);

    return {
      profile: {
        name: profile.name || '安鼎禹',
        title: profile.title || '',
        quote: profile.quote || '',
        avatar: profile.avatar && !profile.avatar.includes('picsum.photos') ? profile.avatar : '/ady/avatar.png',
        about: profile.about || '',
        contact: {
          age: profile.age || 0,
          location: profile.location || '',
          phone: profile.phone || '',
          email: profile.email || ''
        }
      },
      education: education || [],
      workExperience: (workExperience || []).map(exp => ({
        company: exp.company,
        role: exp.role,
        period: exp.period,
        description: exp.description || '',
        details: exp.details || []
      })),
      skillCategories: (skillCategories || []).map(cat => ({
        title: cat.title,
        skills: (cat.skills || []).map(s => ({ name: s.name }))
      })),
      projects: (projects || []).map(proj => ({
        name: proj.name,
        role: proj.role,
        period: proj.period,
        description: proj.description || '',
        details: proj.details || [],
        achievements: proj.achievements || [],
        subProjects: (proj.sub_projects || []).map(sub => ({
          title: sub.title,
          subtitle: sub.subtitle,
          description: sub.description,
          techStack: (sub.tech_stack || []).join(' · ')
        }))
      }))
    };
  } catch (error) {
    console.error('Supabase 数据加载失败:', error);
    throw error;
  }
}

async function generatePDF() {
  const { profile, education, workExperience, skillCategories, projects } = await loadDataFromSupabase();
  
  // 处理头像 URL
  let avatarUrl = profile.avatar;
  if (avatarUrl.startsWith('/')) {
    avatarUrl = 'https://yfarer.cn' + avatarUrl;
  }
  
  // 动态生成个人介绍（与网站一致）
  let dynamicAbout = profile.about || '';
  
  // 获取最新的技能
  const skillNames = skillCategories?.flatMap(cat => 
    cat.skills?.slice(0, 3).map(s => s.name)
  ) || [];
  
  // 获取最新的项目
  const latestProjects = projects?.slice(0, 2).map(p => p.name) || [];
  
  // 添加技能描述
  if (skillNames.length > 0) {
    dynamicAbout += ` 目前专注于${skillNames.slice(0, 3).join('、')}等技术领域。`;
  }
  
  // 添加项目描述
  if (latestProjects.length > 0) {
    dynamicAbout += ` 近期正在开发${latestProjects.join('、')}等项目。`;
  }
  
  // 添加笔记描述（从数据库查询笔记数量）
  try {
    const { data: notes } = await supabase
      .from('notes')
      .select('id')
      .eq('status', 'published');
    const notesCount = notes?.length || 0;
    if (notesCount > 0) {
      dynamicAbout += ` 持续学习并记录了${notesCount}篇技术笔记。`;
    }
  } catch (e) {
    // 忽略笔记查询错误
  }
  
  const templateData = {
    name: profile.name,
    avatar: avatarUrl,
    title: profile.title,
    quote: profile.quote,
    age: profile.contact?.age,
    location: profile.contact?.location,
    phone: profile.contact?.phone,
    email: profile.contact?.email,
    about: dynamicAbout,
    skillCategories: skillCategories,
    projects: projects.map(proj => ({
      name: proj.name,
      role: proj.role,
      period: proj.period,
      description: Array.isArray(proj.description) 
        ? proj.description.map((item, index) => `${index + 1}. ${item}`).join('\n')
        : proj.description,
      details: proj.details,
      achievements: proj.achievements,
      subProjects: proj.subProjects
    })),
    experiences: workExperience.map(exp => ({
      company: exp.company,
      role: exp.role,
      period: exp.period,
      description: exp.description,
      details: exp.details.length > 0 ? exp.details : exp.description.split('；').filter(d => d.trim())
    })),
    education: education
  };
  
  Handlebars.registerHelper('add', function(a, b) {
    return a + b;
  });
  
  Handlebars.registerHelper('breaklines', function(text) {
    if (!text) return '';
    return new Handlebars.SafeString(text.replace(/\n/g, '<br>'));
  });
  
  const template = Handlebars.compile(fs.readFileSync(TEMPLATE_PATH, 'utf8'));
  const html = template(templateData);

  const browser = await puppeteer.launch({ 
    headless: true, 
    executablePath: '/usr/bin/google-chrome',
    args: ['--no-sandbox', '--disable-setuid-sandbox'] 
  });
  
  const page = await browser.newPage();
  await page.setContent(html, { waitUntil: 'networkidle0' });
  
  const height = await page.evaluate(() => document.querySelector('.resume').scrollHeight);
  const heightInMm = Math.ceil(height * 0.264583);
  
  const pdf = await page.pdf({
    width: '210mm',
    height: `${heightInMm}mm`,
    printBackground: true
  });

  await browser.close();
  return pdf;
}

app.get('/api/resume/pdf', async (req, res) => {
  try {
    const pdf = await generatePDF();
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', 'inline; filename="resume.pdf"');
    res.send(pdf);
  } catch (error) {
    console.error('PDF Error:', error);
    res.status(500).json({ error: 'PDF生成失败', msg: error.message });
  }
});

app.get('/api/health', (req, res) => res.json({ status: 'ok', service: 'resume-pdf-api', source: 'supabase' }));

app.listen(PORT, () => console.log(`PDF API on port ${PORT} (Supabase source)`));

File: api/pdf_server.py
from flask import Flask, send_file, jsonify
import json
import os
import base64
import requests
from playwright.sync_api import sync_playwright
from PIL import Image
import io

app = Flask(__name__)

# 数据目录
DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')
TEMPLATE_PATH = os.path.join(os.path.dirname(__file__), '..', 'templates', 'resume-extended.html')

def load_data():
    """加载所有数据文件"""
    with open(os.path.join(DATA_DIR, 'profile.json'), 'r', encoding='utf-8') as f:
        profile = json.load(f)
    with open(os.path.join(DATA_DIR, 'skills.json'), 'r', encoding='utf-8') as f:
        skills = json.load(f)
    with open(os.path.join(DATA_DIR, 'experience.json'), 'r', encoding='utf-8') as f:
        experience = json.load(f)
    with open(os.path.join(DATA_DIR, 'projects.json'), 'r', encoding='utf-8') as f:
        projects = json.load(f)
    with open(os.path.join(DATA_DIR, 'education.json'), 'r', encoding='utf-8') as f:
        education = json.load(f)
    return profile, skills, experience, projects, education

def image_to_base64(url):
    """下载图片并转为base64"""
    try:
        if not url.startswith('http'):
            url = 'https://yfarer.cn' + url
        response = requests.get(url, timeout=10)
        if response.status_code == 200:
            ext = os.path.splitext(url)[1].lower()
            mime_type = 'image/png' if ext == '.png' else 'image/jpeg'
            return f"data:{mime_type};base64,{base64.b64encode(response.content).decode()}"
    except Exception as e:
        print(f"图片下载失败: {e}")
    return None

def prepare_template_data():
    """准备模板数据"""
    profile, skills, experience, projects, education = load_data()
    
    # 处理技能
    skill_categories = [{
        'title': cat['category'],
        'skills': [{'name': item} for item in cat['items']]
    } for cat in skills['skills']['technical']]
    
    # 处理项目
    processed_projects = []
    for proj in projects['projects']:
        sub_projects = None
        if proj.get('subProjects'):
            sub_projects = [{
                'title': sub['title'],
                'subtitle': sub.get('subtitle', ''),
                'description': sub['description'],
                'techStack': ', '.join(sub['techStack']) if sub.get('techStack') else None
            } for sub in proj['subProjects']]
        
        processed_projects.append({
            'name': proj['name'],
            'role': proj['role'],
            'period': proj['period'],
            'tags': proj.get('tags', []),
            'description': proj['description'][0] if isinstance(proj.get('description'), list) else proj.get('description'),
            'subProjects': sub_projects
        })
    
    # 处理工作经历
    processed_experiences = [{
        'company': exp['company'],
        'role': exp['position'],
        'period': exp['period'],
        'description': exp['description'],
        'achievements': exp.get('achievements', [])
    } for exp in experience['experiences']]
    
    # 处理头像
    avatar_base64 = image_to_base64(profile['avatar'])
    
    return {
        'name': profile['name'],
        'avatar': avatar_base64 or profile['avatar'],
        'title': profile['title'],
        'quote': profile['quote'],
        'age': profile['contact']['age'],
        'location': profile['contact']['location'],
        'phone': profile['contact']['phone'],
        'email': profile['contact']['email'],
        'about': profile['about'],
        'skillCategories': skill_categories,
        'projects': processed_projects,
        'experiences': processed_experiences,
        'education': education['education']
    }

def generate_html():
    """生成HTML"""
    from jinja2 import Template
    
    data = prepare_template_data()
    
    with open(TEMPLATE_PATH, 'r', encoding='utf-8') as f:
        template_content = f.read()
    
    template = Template(template_content)
    return template.render(**data)

def generate_pdf():
    """使用Playwright生成PDF"""
    html_content = generate_html()
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True, args=['--no-sandbox'])
        page = browser.new_page()
        
        # 加载HTML
        page.set_content(html_content, wait_until='networkidle')
        page.wait_for_timeout(500)
        
        # 获取页面高度
        body_height = page.evaluate('() => document.body.scrollHeight')
        
        # 设置视口
        page.set_viewport_size({'width': 794, 'height': body_height})
        
        # 截图
        screenshot = page.screenshot(full_page=True, type='png')
        browser.close()
    
    # 转换为PDF
    image = Image.open(io.BytesIO(screenshot))
    
    # 计算PDF尺寸 (A4宽度595.28 points)
    pdf_width = 595.28
    pdf_height = (image.height / image.width) * pdf_width
    
    # 创建PDF
    pdf_buffer = io.BytesIO()
    image.save(pdf_buffer, format='PDF', resolution=100.0)
    pdf_buffer.seek(0)
    
    return pdf_buffer

@app.route('/api/resume/pdf')
def get_resume_pdf():
    """生成并返回简历PDF"""
    try:
        print("PDF请求 received")
        pdf_buffer = generate_pdf()
        print(f"PDF生成完成，大小: {pdf_buffer.getbuffer().nbytes}")
        
        return send_file(
            pdf_buffer,
            mimetype='application/pdf',
            as_attachment=False,
            download_name='resume.pdf'
        )
    except Exception as e:
        print(f"PDF生成失败: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': 'PDF生成失败', 'message': str(e)}), 500

@app.route('/api/health')
def health():
    """健康检查"""
    return jsonify({'status': 'ok', 'service': 'resume-pdf-api-python'})

if __name__ == '__main__':
    print("简历PDF API服务启动在端口 3002")
    print("API地址: http://localhost:3002/api/resume/pdf")
    app.run(host='0.0.0.0', port=3002, debug=False)
